[naga hlsl-out] HLSL implemention of external textures

jamienicol opened the pull request:
**Connections**
Depends on https://github.com/gfx-rs/wgpu/pull/7822
Part of https://github.com/gfx-rs/wgpu/issues/4386

**Description**
HLSL implementation of external textures

**Testing**
Snapshot tests. Manual testing.

**Squash or Rebase?**
Rebase after manually squashing any `fixup` commits that arise from review

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [ ] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

Still not changelog worthy but getting close!

jamienicol committed:
[naga hlsl-out] Implement external texture support

This adds HLSL backend support for `ImageClass::External` (ie WGSL's
`external_texture` texture type).

For each external texture global variable in the IR, we declare 3
`Texture2D` globals as well as a `cbuffer` for the params. The
additional bindings required by these are found in the newly added
`external_texture_binding_map`. Unique names for each can be obtained
using `NameKey::ExternalTextureGlobalVariable`.

For functions that contain ImageQuery::Size, ImageLoad, or ImageSample
expressions for external textures, ensure we have generated wrapper
functions for those expressions. When emitting code for the
expressions themselves, simply insert a call to the wrapper function.

For size queries, we return the value provided in the params
struct. If that value is [0, 0] then we query the size of the plane 0
texture and return that.

For load and sample, we sample the textures based on the number of
planes specified in the params struct. If there is more than one plane
we additionally perform YUV to RGB conversion using the provided
matrix.

Unfortunately HLSL does not allow structs to contain textures, meaning
we are unable to wrap the 3 textures and params struct variables in a
single variable that can be passed around.

For our wrapper functions we therefore ensure they take the three
textures and the params as consecutive arguments. Likewise, when
declaring user-defined functions with external texture arguments, we
expand the single external texture argument into 4 consecutive
arguments. (Using NameKey::ExternalTextureFunctionArgument to ensure
unique names for each.)

Thankfully external textures can only be used as either global
variables or function arguments. This means we only have to handle the
`Expression::GlobalVariable` and `Expression::FunctionArgument` cases
of `write_expr()`. Since in both cases we know the external texture
can only be an argument to either a user-defined function or one of
our wrapper functions, we can simply emit the names of the variables
for each three textures and the params struct in a comma-separated
list.

jamienicol committed:
[naga] Reserve names for each plane and params buffer of external texture

Adds new `NameKey` variants `ExternalTextureGlobalVariable` and
`ExternalTextureFunctionArgument`, like their non-external-texture
cousins but additionally keyed by either being a specific plane index
or params buffer.

For each external texture global variable or function argument reserve
additional names for 3 planes and the params buffer. For Naga backends
which must represent external textures as multiple variables/arguments,
this will allow them to uniquely name each one.

jamienicol committed:
[naga] Generate special type for external texture params buffer

During wgsl lowering, if we encounter an external texture type then
generate the `ExternalTextureParams` struct. This will be required by
most Naga backends to implement external textures.

This type is not actually used by wgsl-in or the IR. However,
generating it in Naga IR ensures tricky details such as member
alignment are handled for us.

wgsl-out must ensure it does *not* generate code for this type, as it
handles external textures natively.

jamienicol commented:
Rebased and updated `textureLoad` implementation to use `load_transform` parameter, as described [here](https://github.com/gfx-rs/wgpu/pull/7823#issuecomment-3013049539)

jimblandy commented:
Rebased on trunk since #7822 is landed.

jimblandy commented:
Rebased on trunk since #7823 has landed, and regenerated snapshot output.

jimblandy commented:
Not a request for a change, just some thoughts:

Does `NagaExternalTextureParams` really need to be in the IR? It's pretty strange to have a type in there that the code never uses, and that is used only by some backends and must be ignored by others. This seems like something that's going to break easily or confuse people, because it certainly isn't a typical use of IR types. For example, it might seem reasonable to someone that compaction should not treat special types as axiomatically rooted - but changing compaction to remove apparently unused special types would break whoever's using `external_texture_params`.

In the long term, I'm wondering if Naga doesn't need two IRs, one designed around front-end needs, and a second around backend needs, with a lowering pass between them. The lowering code could be shared by backends, but it would take parameters indicating which transformations the backends needed. One such transformation might be, "lower external textures to ordinary textures and parameter blocks". Others might be the HLSL access rewriting, or the unrestricted pointer parameters rewrites.

jimblandy commented:
@jamienicol Do the doc changes in <del>[7741fba](https://github.com/gfx-rs/wgpu/pull/7826/commits/7741fbaf578e986ac5f567d0b322e64bba9b931d)</del> [611c3c9](https://github.com/gfx-rs/wgpu/pull/7826/commits/611c3c998ca64c8bee593607173f6cb97a4fc78b) look correct?

jimblandy commented on naga/src/back/wgsl/writer.rs:
```diff
@@ -109,12 +109,28 @@ impl<W: Write> Writer<W> {
             .values()
-            .any(|t| *t == handle)
+            .any(|t| *t == ty)
+            || Some(ty) == module.special_types.external_texture_params
```
Again, not requesting a change, but this is weird. People will wonder why all the other `special_types` aren't listed here too.

jamienicol replied:
Perhaps the other `special_types` should in fact be listed here too? Looking at the input wgsl snapshot tests, eg [here](https://github.com/gfx-rs/wgpu/blob/trunk/naga/tests/in/wgsl/ray-query.wgsl#L19), the special types aren't actually declared in the source (they are in a comment just to show what they would look like). The respective TOML files specify not to run WGSL-out on them, but presumably if they did we would indeed want to skip emitting the declarations. Otherwise when parsing again with wgsl-in we would have duplicated declarations?

jimblandy replied:
Okay, filed as #8026.

jimblandy reviewed (commented):
The first commit, "[naga] Generate special type for external texture params buffer", looks good to me.

jimblandy commented:
@jamienicol In [17f8bf8](https://github.com/gfx-rs/wgpu/pull/7826/commits/17f8bf89b44cebb9df56b1200c6bf2f6b762ed2f) I was trying to reduce the repetition in `Namer::reset`, but I'm not sure I succeeded. If it doesn't seem like a clear improvement to you, we can drop that `fixup!`.

jimblandy reviewed (commented):
The second commit, "[naga] Reserve names for each plane and params buffer of external texture", looks good to me.

jimblandy reviewed (changes requested):
The log message for the third commit - the long one - is superb, but it belongs in the code, not in the commit message.

jimblandy commented:
Sorry - still reviewing that final commit.

jamienicol commented:
> @jamienicol Do the doc changes in ~[7741fba](https://github.com/gfx-rs/wgpu/pull/7826/commits/7741fbaf578e986ac5f567d0b322e64bba9b931d)~ [611c3c9](https://github.com/gfx-rs/wgpu/pull/7826/commits/611c3c998ca64c8bee593607173f6cb97a4fc78b) look correct?

They look great, modulo a typo which I've fixed in 5f90c8e2

jamienicol commented:
> The log message for the third commit - the long one - is superb, but it belongs in the code, not in the commit message.

I believe the important contents of it is all in the code, spread around various places as appropriate. But a single high-level overview of how external textures are implemented for HLSL in the code somewhere doesn't sound like a bad thing. Any suggestions for where to put this?

jamienicol commented:
> Does `NagaExternalTextureParams` really need to be in the IR? It's pretty strange to have a type in there that the code never uses, and that is used only by some backends and must be ignored by others. This seems like something that's going to break easily or confuse people, because it certainly isn't a typical use of IR types. For example, it might seem reasonable to someone that compaction should not treat special types as axiomatically rooted - but changing compaction to remove apparently unused special types would break whoever's using `external_texture_params`.

No, it doesn't _need_ to be in the IR. The benefit of it being there is that we don't have to duplicate its declaration in each backend, and it takes care of fiddly bits such as adding padding, packing a `mat3x2` as 3x `vec2` in HLSL, etc, for us. When I stumbled across the existing `special_types` for ray queries etc it seemed conceptually similar-ish to me. 

(You're right to point out a confusing difference in how the code as-is handles my params struct vs existing special types, which I've replied to above.)

I'm afraid I don't know what "axiomatically rooted" means. But FWIW our tests would explode if somebody made that change to compaction and it would be quickly caught in CI.

None of this is required, and if you think the weirdness outweighs the benefits I am more than happy to emit the declarations in the backends instead.

jamienicol commented:
Also pushed 84d4f0ea which makes your `as_deref()` change for `NameKey::ExternalTextureFunctionArgument` to `NameKey::ExternalTextureGlobalVariable` as well.

jamienicol commented:
Pushed [678de10](https://github.com/gfx-rs/wgpu/pull/7826/commits/678de1006074d9d7f0957096e2bafb1bf7dc9d83) to add a comment saying why we don't want to compact away the params special type

And [d5a4687](https://github.com/gfx-rs/wgpu/commit/d5a4687e92af79c783084c45f797227c819ce568) adding a module-level comment

jimblandy commented:
Hmm. This doesn't properly ensure that `NagaExternalDimensions2D` is a unique name. But this is a pre-existing issue for all query functions. Filed as #8025.

jimblandy reviewed (commented):
It seems to me that `sample_transform` must be a transform that takes the (0,0) - (1,1) square to an axis-aligned rectangle that lies within (0,0) - (1,1), because otherwise the use of `min` and `max` to compute `bounds` in `nagaTextureSampleBaseClampToEdge` doesn't work.

If that's correct, could we get that commented? Perhaps the doc comment for `SpecialTypes::external_texture_params` would be a good place.

jimblandy commented:
Is it really the case that `nagaTextureSampleBaseClampToEdge` doesn't need to take `params.size` into account, but `nagaTextureLoadExternal` does? If so, that deserves a comment.

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -254,12 +264,233 @@ impl<W: Write> super::Writer<'_, W> {
+                    self.out,
+                    "{l3}plane2.GetDimensions(plane2_size.x, plane2_size.y);"
+                )?;
+                writeln!(self.out, "{l2}uint2 plane2_coords = uint2(floor(float2(plane0_coords) * float2(plane2_size) / float2(plane0_size)));")?;
```
This should be `{l3}`, right?

jamienicol replied:
correct

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -254,12 +264,233 @@ impl<W: Write> super::Writer<'_, W> {
+                // If params.size is zero then clamp to the actual size of the texture.
+                writeln!(
+                    self.out,
+                    "{l1}uint2 cropped_size = params.size != 0 ? params.size : plane0_size;"
```
Is there a reason we say `params.size != 0` here but `any(params.size)` in the query? If so, it needs a comment.

jamienicol replied:
No reason. You have to use `any()` in the query as the `!=` evaluates to a vec, which cannot be used as an `if` condition. Whereas here it lets me use a vec for a _component wise_ ternary. But that's really not what we want. We want to use a scalar condition so I'll change this to use `any` too.

jimblandy commented:
@jamienicol I think Naga does support `TypeInner::BindingArray`s of images, but this PR's approach doesn't support a `BindingArray` of `External` images. The validator should forbid that.

I should have caught this in the review for #7822, sorry about that.

jimblandy reviewed (approved):
All right! Just some minor comments.

jamienicol commented:
> It seems to me that `sample_transform` must be a transform that takes the (0,0) - (1,1) square to an axis-aligned rectangle that lies within (0,0) - (1,1), because otherwise the use of `min` and `max` to compute `bounds` in `nagaTextureSampleBaseClampToEdge` doesn't work.
> 
> If that's correct, could we get that commented? Perhaps the doc comment for `SpecialTypes::external_texture_params` would be a good place.

Yes this is correct. I think `wgpu_core::device::resource::ExternalTextureParams` would be the better place to document this, and I'll link to there from `SpecialTypes::external_texture_params`. Unfortunately it can't be a _real_ link because they're in different crates, but I still think that's the most sensible place, as if a field is added/removed/changed in that struct the documentation is right there.

> Is it really the case that `nagaTextureSampleBaseClampToEdge` doesn't need to take `params.size` into account, but `nagaTextureLoadExternal` does? If so, that deserves a comment.

This is really the case. `textureSampleBaseClampToEdge()` works in normalized coordinates so the texture size doesn't matter, we just need to clamp to 0..1. (with transform and half-texel adjustment). `textureLoad` works with unnormalized coordinates so we need to clamp to the size the external texture claims to be. (not the actual texture size, as there may be a crop rect)

Perhaps we could do the equivalent of what we do in the sample case, eg transform (0,0)..(texture_size - 1) and then clamp within that rect. But that's more instructions, and I seem to recall there was an issue with that approach anyway although unfortunately I can no longer remember the specifics

> @jamienicol I think Naga does support `TypeInner::BindingArray`s of images, but this PR's approach doesn't support a `BindingArray` of `External` images. The validator should forbid that.
> 
> I should have caught this in the review for #7822, sorry about that.

I'll see if I can fix this

jamienicol commented:
I filed https://github.com/gfx-rs/wgpu/issues/8027 for actually implementing binding array support

jamienicol committed:
[naga] Have validation reject shaders containing binding arrays of external textures

For simplicity's sake our initial implementation of external textures
will not support binding arrays of external textures. We should
therefore reject any shaders which use them during validation.

Their implementation will be tracked in #8027.

naga/src/valid/type.rs JJ: JJ: Lines starting with "JJ:" (like this
one) will be removed.

jimblandy commented:
@jamienicol Okay, I pushed one more doc tweak, but this looks good to go. The binding array check looked right to me.

Thanks!!
